packages/realm/src/app-services/BaseSubscriptionSet.ts

```diff
@@ -0,0 +1,216 @@
+   * that someone has called {@link SubscriptionSet.update} on a different instance
+   * of the `Subscriptions`. You should not use a superseded SubscriptionSet,
+   * and instead obtain a new instance from {@link Realm.subscriptions}.
+   */
+  Superseded = "superseded",
+}
+
+const DEFAULT_PROPERTY_DESCRIPTOR: PropertyDescriptor = { configurable: true, enumerable: true, writable: true };
+const PROXY_HANDLER: ProxyHandler<BaseSubscriptionSet> = {
+  get(target, prop) {
+    if (Reflect.has(target, prop)) {
+      return Reflect.get(target, prop);
+    }
+    if (typeof prop === "string") {
+      const BASE = 10;
+      const index = Number.parseInt(prop, BASE);
+      // TODO: Consider catching an error from access out of bounds, instead of checking the length, to optimize for the hot path
+      if (!Number.isNaN(index) && index >= 0 && index < target.length) {
+        return target.get(index);
+      }
+    }
+  },
+  ownKeys(target) {
+    return Reflect.ownKeys(target).concat([...target.keys()].map(String));
+  },
+  getOwnPropertyDescriptor(target, prop) {
+    if (Reflect.has(target, prop)) {
+      return Reflect.getOwnPropertyDescriptor(target, prop);
+    } else if (typeof prop === "string") {
+      const BASE = 10;
+      const index = Number.parseInt(prop, BASE);
+      if (index < target.length) {
+        return DEFAULT_PROPERTY_DESCRIPTOR;
+      }
+    }
+  },
+};
+
+/**
+ * Class representing the common functionality for the {@link SubscriptionSet} and
+ * {@link MutableSubscriptionSet} classes.
+ *
+ * The {@link Subscription}s in a SubscriptionSet can be accessed as an array, e.g.
+ * `realm.subscriptions[0]`. This array is readonly – SubscriptionSets can only be
+ * modified inside a {@link SubscriptionSet.update} callback.
+ */
+export abstract class BaseSubscriptionSet {
+  // Enables index accessing when combined with the proxy handler's `get()` method
+  readonly [n: number]: Subscription;
+
+  /**@internal */
+  protected constructor(/**@internal */ protected internal: binding.SyncSubscriptionSet) {
+    return new Proxy(this, PROXY_HANDLER);
+  }
+
+  /**
+   * Determine whether the SubscriptionSet is empty.
+   *
+   * @returns `true` if there are no subscriptions in the set, otherwise `false`.
+   */
+  get isEmpty(): boolean {
+    return this.internal.size === 0;
+  }
+
+  /**
+   * Get the version of the SubscriptionSet.
+   *
+   * @returns The version of the SubscriptionSet. This is incremented every time an
+   *  {@link SubscriptionSet.update} is applied.
+   */
+  get version(): number {
+    return Number(this.internal.version);
+  }
+
+  /**
+   * Get the state of the SubscriptionSet.
+   *
+   * @returns The state of the SubscriptionSet.
+   */
+  get state(): SubscriptionsState {
+    const stateValue = this.internal.state.valueOf();
+    switch (stateValue) {
+      case 0: // Uncommitted
+      case 1: // Pending
+      case 2: // Bootstrapping
+      case 6: // AwaitingMark
```

Comment:
This case was not used in the previous implementation, meaning it would have thrown an error if `AwaitingMark` was the state. Should we treat this as `Pending`?

Reply:
I think you are right, and it will align us the [Realm Swift](https://github.com/realm/realm-swift/blob/master/Realm/RLMSyncSubscription.mm#L219)

Reply:
Please resolve the conversation as we agree that case `6` should be included :smile: